src/combinatorics/ballot.lean

```diff
@@ -0,0 +1,453 @@
+/-
+Copyright (c) 2022 Bhavik Mehta, Kexing Ying. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta, Kexing Ying
+-/
+import probability.cond_count
+
+/-!
+# Ballot problem
+
+This file proves Theorem 30 from the [100 Theorems List](https://www.cs.ru.nl/~freek/100/).
+
+The ballot problem asks, if in an election, candidate A receives `p` votes whereas candidate B
+receives `q` votes where `p > q`, what is the probability that candidate A is strictly ahead
+throughout the count. The probability of this is `(p - q) / (p + q)`.
+
+## Main definitions
+
+* `counted_sequence`: given natural numbers `p` and `q`, `counted_sequence p q` is the set of
+  all lists containing `p` of `1`s and `q` of `-1`s representing the votes of candidate A and B
+  respectively.
+* `stays_positive`: is the set of lists of integers which suffix has positive sum. In particular,
+  the intersection of this set with `counted_sequence` is the set of lists where candidate A is
+  strictly ahead.
+
+## Main result
+
+* `ballot`: the ballot problem.
+
+-/
+
+open set probability_theory measure_theory
+
+/-- The set of nonempty lists of integers which suffix has positive sum. -/
+def stays_positive : set (list ℤ) := {l | ∀ l₂, l₂ ≠ [] → l₂ <:+ l → 0 < l₂.sum}
+
+@[simp] lemma stays_positive_nil : [] ∈ stays_positive :=
+λ l hl hl₁, (hl (list.eq_nil_of_suffix_nil hl₁)).elim
+
+lemma stays_positive_cons_pos (x : ℤ) (hx : 0 < x) (l : list ℤ) :
+  (x :: l) ∈ stays_positive ↔ l ∈ stays_positive :=
+begin
+  split,
+  { intros hl l₁ hl₁ hl₂,
+    apply hl l₁ hl₁ (hl₂.trans (list.suffix_cons _ _)) },
+  { intros hl l₁ hl₁ hl₂,
+    rw list.suffix_cons_iff at hl₂,
+    rcases hl₂ with (rfl | hl₂),
+    { rw list.sum_cons,
+      apply add_pos_of_pos_of_nonneg hx,
+      cases l with hd tl,
+      { simp },
+      { apply le_of_lt (hl (hd :: tl) (list.cons_ne_nil hd tl) (hd :: tl).suffix_refl) } },
+    { apply hl _ hl₁ hl₂ } }
+end
+
+@[elab_as_eliminator]
+lemma diag_induction (P : ℕ → ℕ → Prop) (ha : ∀ a, P (a + 1) (a + 1)) (hb : ∀ p, P 0 (p + 1))
```

Comment:
This lemma should be near `nat.le_induction` and it should get a proper docstring.